subsys/settings/src/settings.c

```diff
@@ -32,27 +30,92 @@ void settings_init(void)
 
 int settings_register(struct settings_handler *handler)
 {
-	if (settings_handler_lookup(handler->name)) {
-		return -EEXIST;
+	struct settings_handler *ch;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&settings_handlers, ch, node) {
+		/* avoid registering a setting handler that is sub or a super of
+		 * existing handler
+		 */
+		if ((settings_name_cmp(handler->name, ch->name, NULL) == 1) ||
+		    (settings_name_cmp(ch->name, handler->name, NULL) == 1)) {
+			return -EINVAL;
+		}
+		/* avoid registering existing handlers */
+		if (settings_name_cmp(handler->name, ch->name, NULL) == 0) {
+			return -EEXIST;
+		}
+
 	}
 	sys_slist_prepend(&settings_handlers, &handler->node);
 
 	return 0;
 }
 
-/*
- * Find settings_handler based on name.
- */
-static struct settings_handler *settings_handler_lookup(char *name)
+int settings_name_cmp(const char *name, const char *key, const char **next)
 {
-	struct settings_handler *ch;
+	if (next) {
+		*next = NULL;
+	}
 
-	SYS_SLIST_FOR_EACH_CONTAINER(&settings_handlers, ch, node) {
-		if (!strcmp(name, ch->name)) {
-			return ch;
+	/* name might come from flash directly, in flash the name would end
+	 * with '=' or '\0' depending how storage is done. Flash reading is
```

Comment:
why '=' is possible?

Reply:
When you are directly reading from flash the name ends with `=`.

Reply:
Agree - this is something I was planning to implement when we provide direct access from FLASH in the backends where it is possible. But it is a good option to have it even now.